src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp

```diff
@@ -241,6 +224,262 @@ void C2_MacroAssembler::fast_unlock(Register objectReg, Register boxReg, Registe
+  br(Assembler::NE, flag_correct);
+  stop("Fast Lock Flag != NE");
+  bind(flag_correct);
+#endif
+  // C2 uses the value of Flags (NE vs EQ) to determine the continuation.
+}
+
+void C2_MacroAssembler::fast_unlock_lightweight(Register obj, Register t1, Register t2,
+                                                Register t3) {
+  assert(LockingMode == LM_LIGHTWEIGHT, "must be");
+  assert_different_registers(obj, t1, t2, t3);
+
+  // Handle inflated monitor.
+  Label inflated, inflated_load_monitor;
+  // Finish fast unlock successfully. MUST branch to with flag == EQ
+  Label unlocked;
+  // Finish fast unlock unsuccessfully. MUST branch to with flag == NE
+  Label slow_path;
+
+  const Register mark = t1;
+  const Register top = t2;
+  const Register t = t3;
+
+  { // Lightweight unlock
+
+    // Check if obj is top of lock-stack.
+    ldrw(top, Address(rthread, JavaThread::lock_stack_top_offset()));
+    subw(top, top, oopSize);
+    ldr(t, Address(rthread, top));
+    cmp(obj, t);
+    // Top of lock stack was not obj. Must be monitor.
+    br(Assembler::NE, inflated_load_monitor);
+
+    // Pop lock-stack.
+    DEBUG_ONLY(str(zr, Address(rthread, top));)
+    strw(top, Address(rthread, JavaThread::lock_stack_top_offset()));
+
+    // Check if recursive.
+    subw(t, top, oopSize);
+    ldr(t, Address(rthread, t));
+    cmp(obj, t);
+    br(Assembler::EQ, unlocked);
+
+    // Not recursive.
+    // Load Mark.
+    ldr(mark, Address(obj, oopDesc::mark_offset_in_bytes()));
+
+    // Check header for monitor (0b10).
+    tbnz(mark, exact_log2(markWord::monitor_value), inflated);
+
+    // Try to unlock. Transition lock bits 0b00 => 0b01
+    assert(oopDesc::mark_offset_in_bytes() == 0, "required to avoid lea");
+    orr(t, mark, markWord::unlocked_value);
+    cmpxchg(/*addr*/ obj, /*expected*/ mark, /*new*/ t, Assembler::xword,
+            /*acquire*/ false, /*release*/ true, /*weak*/ false, noreg);
+    br(Assembler::EQ, unlocked);
+
+    // Load link store conditional exclusive failed.
+    // Restore lock-stack and handle the unlock in runtime.
+    DEBUG_ONLY(str(obj, Address(rthread, top));)
+    addw(top, top, oopSize);
+    str(top, Address(rthread, JavaThread::lock_stack_top_offset()));
+    b(slow_path);
+  }
+
+
+  { // Handle inflated monitor.
+    bind(inflated_load_monitor);
+    ldr(mark, Address(obj, oopDesc::mark_offset_in_bytes()));
+#ifdef ASSERT
+    tbnz(mark, exact_log2(markWord::monitor_value), inflated);
+    stop("Fast Unlock not monitor");
+#endif
+
+    bind(inflated);
+
+#ifdef ASSERT
+    Label check_done;
+    subw(top, top, oopSize);
+    cmpw(top, in_bytes(JavaThread::lock_stack_base_offset()));
+    br(Assembler::LT, check_done);
+    ldr(t, Address(rthread, top));
+    cmp(obj, t);
+    br(Assembler::NE, inflated);
+    stop("Fast Unlock lock on stack");
+    bind(check_done);
+#endif
+
+    // mark contains the tagged ObjectMonitor*.
+    const Register monitor = mark;
```

Comment:
I wouldn't do this. Making `monitor` a register alias for `mark` is at best confusing and at worst dangerous. As a general rule, aliased register names have caused many bugs.

Reply:
I have been trying to think of alternatives which improve this situation. 

Is it the fact that monitor alias mark, or any of the aliases used for the in-parameters that is confusing?
If it is the first do you see the following as a solution:
```c++
    const Register monitor = t1;
    const uintptr_t monitor_tag = markWord::monitor_value;

    // Untag the monitor.
    sub(monitor, mark, monitor_tag);
```
and for the lock case
```c++
    const Register tagged_monitor = t1;
    mov(tagged_monitor, mark); 
```
where the mov will emit no instruction as they are the same register.
(These two variants seem somewhat equivalent to me, but it was not clear from your comment if this is the issue)

If it is the second, I feel like the best compromise would to still alias register via named variables, but give them names that show their origin (so you do not need a language server or scroll/grep to get the contextual information). 
E.g.
```c++
    const Register t1_monitor = t1;
    const uintptr_t monitor_tag = markWord::monitor_value;

    // Untag the monitor.
    sub(t1_monitor, t1_mark, monitor_tag);
```

I would like to hear if I have misunderstood something and / or if you have another proposed solution to this issue.

This seem to be the final issue blocking integration.

Reply:
These are both misleading to the reader. Sure, they're cute, but they're dangerous, as has proved in practice several times in the AArch64 port. Writing a `mov` that is known to do nothing is severely misleading. It can be mitigated with a "this mov does nothing" comment.

A prefix naming convention is a significant improvement, . It may also be a good idea to poison the no-longer-valid names (with a `#define`).

Bear in mind that the typical reader is skim-reading this stuff.

I've been trying a better way to do register allocation in hand-coded assembly, where a `RegSet` contains the currently-free registers. At the end of its live range, a value is pushed back into the free set. This avoids most of the pitfalls. (I'm not suggesting that here, it's something to think about for tr future.)

Reply:
> These are both misleading to the reader. Sure, they're cute, but they're dangerous, as has proved in practice several times in the AArch64 port. Writing a `mov` that is known to do nothing is severely misleading. It can be mitigated with a "this mov does nothing" comment.

I suspected that this interpretation of what you was asking was incorrect, but wanted to clarify. 

> A prefix naming convention is a significant improvement

I will go ahead with this approach then.